### PR TITLE
Fixes Validation::Configure ignore failures bug

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
Previously the `validate` method on `Validation::Configure` would return an object which contains if the validation succeeded or not. However this was not the object that `success?` used. Instead a stale instance variable was used and thus the validation could fail but still return true. This PR fixes the bug by saving the return object of `validate` as the instance variable.

Also added additional validation checks to allow empty strings as a valid default and to allow `Boolean` default values. Previously these caused the validation to fail but it wasn't being detected due to the previously mentioned bug.

### CODE REVIEW

File: spec/validation/configure_spec.rb [hunk 1]
```diff
@@ -43,6 +43,11 @@
           type: 'integer',
           default: 10,
         },
+        boolean_true: {
+          question: 'Can I have a boolean true (/yes) default?',
+          type: 'boolean',
+          default: 'yes',
```
Comment (bobwhitelock) [inline spec/validation/configure_spec.rb]: I don't think we want to require `boolean` questions have their default specified as a `'yes'` or `'no'` string. Firstly this conversion is already performed within the `Configurator`, so requiring this here is unnecessary and also won't work once the default value is passed to the `Configurator` I believe.

Secondly, the default should be the default value which will be saved if no input is given, via whatever UI the user is using to perform the `configure`; in the case of `boolean` questions this will be `true`/`false` rather than `'yes'`/`'no'`. Doing this makes clear what this value is, and also means we can handle a default from a higher level saved answer file in the same way as a `configure.yaml`-defined default.

The use of `'yes'`/`'no'` to enter a default on the CLI can be considered a detail of this particular interface, used because accepting text is the simplest/standard way to get user input in a CLI. Other interfaces could accept different input to set a `boolean` value however, e.g. the GUI could use a check box.

Can we change this validation to validate for `true`/`false`?

File: src/validation/configure.rb [hunk 2]
```diff
@@ -41,31 +41,32 @@ def initialize(file)
...
           end
+          configure_results
         end
-        @configure_results
       end
 
       def success?
-        validate if @configure_results.nil?
-        @configure_results.success?
+        validate.success?
```
Comment (bobwhitelock) [inline src/validation/configure.rb]: You could remove this method entirely and replace it with `delegate :success?, to: :validate` at the top of this file. You may or may not need to add `require 'active_support/core_ext/module/delegation'` to ensure `delegate` is imported.

File: spec/validation/configure_spec.rb [hunk 3]
```diff
@@ -247,6 +247,17 @@ def run_configure_validation(my_hash = {})
                                   })
       expect(run_configure_validation(h).keys).to eq([:default_integer_type])
     end
+
+    it 'fails if the default is an empty string' do
```
Comment (bobwhitelock) [inline spec/validation/configure_spec.rb]: I don't follow why we need a special case for empty string defaults - can't they be covered by the same validating as non-empty strings? Then as I understand it we wouldn't need this test or special handling for empty strings, as this is just another test for `default_type?`. Am I missing something? This validations stuff still looks like magic to me.

File: src/validation/configure.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply src/validation/configure.rb]: Fixed in: https://github.com/alces-software/metalware/pull/178/commits/f40f29e5897301f20092cb1fe8f4c7afb55978ea

File: spec/validation/configure_spec.rb (hunk 3 above)
Comment (WilliamMcCumstie) [reply spec/validation/configure_spec.rb]: Because of the crazy logic that occurs in the `Validation` I can't guarantee that empty strings will be treated the same way as a regular string. The new `Validator` class in a latter PR handles it better.

File: spec/validation/configure_spec.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply spec/validation/configure_spec.rb]: Will look into this. However it isn't ideal to fix it on this Pr but instead fix the issue off #185

**Comment by WilliamMcCumstie** [comment]
Fixed/ responded to your comments. Some of them might need to be fixed in a latter PR.

This PR has been mostly superseded, but it is the foundation of the refactored code.

**Comment by bobwhitelock** [comment]
Thanks, so long as those 2 comments are noted to be looked at then I'll merge this now.
